mediagrains_py36/numpy/videograin.py

```diff
@@ -0,0 +1,293 @@
+        # This method returns an empty list because component access isn't supported, but
+        # the more basic access to the underlying data is.
+        return []
+
+    raise NotImplementedError("Cog Frame Format not amongst those supported for numpy array interpretation")
+
+
+class VIDEOGRAIN (bytesgrain.VIDEOGRAIN):
+    _grain_conversions = {}
+
+    def __init__(self, meta, data):
+        super().__init__(meta, data)
+        self._data = np.frombuffer(self._data, dtype=_dtype_from_cogframeformat(self.format))
+        self.component_data = ComponentDataList(
+            _component_arrays_for_data_and_type(self._data, self.format, self.components),
+            arrangement=_component_arrangement_from_format(self.format))
+
+    def __array__(self):
+        return np.array(self.data)
+
+    def __bytes__(self):
+        return bytes(self.data)
+
+    def __copy__(self):
+        return VideoGrain(copy(self.meta), self.data)
+
+    def __deepcopy__(self, memo):
+        return VideoGrain(deepcopy(self.meta), self.data.copy())
+
+    def __repr__(self):
+        if self.data is None:
+            return "{}({!r})".format(self._factory, self.meta)
+        else:
+            return "{}({!r},< numpy data of length {} >)".format(self._factory, self.meta, len(self.data))
+
+    @classmethod
+    def grain_conversion(cls, fmt_in: CogFrameFormat, fmt_out: CogFrameFormat):
+        """Decorator to apply to all grain conversion functions"""
+        def _inner(f: Callable[[cls, cls], None]) -> None:
+            cls._grain_conversions[(fmt_in, fmt_out)] = f
+            return f
+        return _inner
+
+    @classmethod
+    def grain_conversion_two_step(cls, fmt_in: CogFrameFormat, fmt_mid: CogFrameFormat, fmt_out: CogFrameFormat):
+        """Register a grain conversion via an intermediate format, using existing conversions"""
+        def _inner(grain_in: "VIDEOGRAIN", grain_out: "VIDEOGRAIN"):
+            grain_mid = grain_in._similar_grain(fmt_mid)
+            cls._get_grain_conversion_function(fmt_in, fmt_mid)(grain_in, grain_mid)
+            cls._get_grain_conversion_function(fmt_mid, fmt_out)(grain_mid, grain_out)
+        cls.grain_conversion(fmt_in, fmt_out)(_inner)
+
+    @classmethod
+    def _get_grain_conversion_function(cls, fmt_in: CogFrameFormat, fmt_out: CogFrameFormat) -> Callable[["VIDEOGRAIN", "VIDEOGRAIN"], None]:
+        """Return the registered grain conversion function for a specified type conversion, or raise NotImplementedError"""
+        if (fmt_in, fmt_out) in cls._grain_conversions:
+            return cls._grain_conversions[(fmt_in, fmt_out)]
+
+        raise NotImplementedError("This conversion has not yet been implemented")
+
+    def flow_id_for_converted_flow(self, fmt: CogFrameFormat) -> uuid.UUID:
+        return uuid.uuid5(self.source_id, "FORMAT_CONVERSION: {!r}".format(fmt))
+
+    def _similar_grain(self, fmt: CogFrameFormat) -> "VIDEOGRAIN":
```

Comment:
Minor: I don't think the return result type annotation requires quotes as VIDEOGRAIN type is known. Another one found in the next function.